src/Servers/Kestrel/Core/src/Internal/Http3/Http3ControlStream.cs

```diff
@@ -0,0 +1,435 @@
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3
+{
+    internal class Http3ControlStream : IHttp3Stream
+    {
+        private Http3FrameWriter _frameWriter;
+        private readonly Http3Connection _http3Connection;
+        private HttpConnectionContext _context;
+        private readonly Http3Frame _incomingFrame = new Http3Frame();
+        private int _isClosed;
+        private int _gracefulCloseInitiator;
+
+        private bool _haveReceivedSettingsFrame;
+
+        public Http3ControlStream(Http3Connection http3Connection, HttpConnectionContext context)
+        {
+            var httpLimits = context.ServiceContext.ServerOptions.Limits;
+
+            _http3Connection = http3Connection;
+            _context = context;
+
+            // Todo framewriter.
+            _frameWriter = new Http3FrameWriter(
+                context.Transport.Output,
+                context.ConnectionContext,
+                this,
+                context.TimeoutControl,
+                httpLimits.MinResponseDataRate,
+                context.ConnectionId,
+                context.MemoryPool,
+                context.ServiceContext.Log);
+            var closedRegistration = _context.ConnectionContext.ConnectionClosed.Register(state => ((Http3ControlStream)state).OnStreamClosed(), this);
+        }
+
+        private void OnStreamClosed()
+        {
+            // TODO how to pass Abort in.
+            Abort(new ConnectionAbortedException("HTTP_CLOSED_CRITICAL_STREAM"));
+        }
+
+        public PipeReader Input => _context.Transport.Input;
+
+
+        public void Abort(ConnectionAbortedException ex)
+        {
+        }
+
+        public void HandleReadDataRateTimeout()
+        {
+            //Log.RequestBodyMinimumDataRateNotSatisfied(ConnectionId, null, Limits.MinRequestBodyDataRate.BytesPerSecond);
+            Abort(new ConnectionAbortedException(CoreStrings.BadRequest_RequestBodyTimeout));
+        }
+
+        public void HandleRequestHeadersTimeout()
+        {
+            //Log.ConnectionBadRequest(ConnectionId, BadHttpRequestException.GetException(RequestRejectionReason.RequestHeadersTimeout));
+            Abort(new ConnectionAbortedException(CoreStrings.BadRequest_RequestHeadersTimeout));
+        }
+
+        public void OnInputOrOutputCompleted()
+        {
+            TryClose();
+            _frameWriter.Abort(new ConnectionAbortedException(CoreStrings.ConnectionAbortedByClient));
+        }
+
+        private bool TryClose()
+        {
+            if (Interlocked.Exchange(ref _isClosed, 1) == 0)
+            {
+                return true;
+            }
+
+            // TODO make this actually close the Http3Stream by telling msquic to close the stream.
+            return false;
+        }
+
+        private async Task HandleEncodingTask()
+        {
+            var encoder = new EncoderStreamReader(10000); // TODO get value from limits
+            while (_isClosed == 0)
```

Comment:
This should be a `Volatile` read, otherwise it won't be guaranteed to observe the write from `Interlocked`. The other reads need this update as well.

```suggestion
            while (Volatile.Read(ref _isClosed) == 0)
```

Reply:
Can I just make the int volatile?

Reply:
yah